src/main/kotlin/org/jitsi/nlj/stats/PacketIOActivity.kt

```diff
@@ -16,14 +16,38 @@
 
 package org.jitsi.nlj.stats
 
-import kotlin.math.max
+import org.jitsi.nlj.util.NEVER
+import org.jitsi.nlj.util.latest
+import java.time.Instant
 
 class PacketIOActivity {
-    var lastPacketReceivedTimestampMs: Long = 0
-    var lastPacketSentTimestampMs: Long = 0
+    var lastRtpPacketReceivedTimestamp: Instant = NEVER
```

Comment:
Also, do they need to be synchronized?

Reply:
I just pushed a change to use `vetoable` to make sure we only go forward.  Unfortunately, from what I can tell, there's no built-in way to synchronize a `vetoable`. The implementation of `vetoable` is pretty simple, though, so we could create our own.  I think it'd be:
```
public inline fun <T> safeVetoable(initialValue: T, crossinline onChange: (property: KProperty<*>, oldValue: T, newValue: T) -> Boolean):
    ReadWriteProperty<Any?, T> =
    object : ObservableProperty<T>(initialValue) {
        private val lock = Any()

        override fun beforeChange(property: KProperty<*>, oldValue: T, newValue: T): Boolean =
            onChange(property, oldValue, newValue)
        
        override fun setValue(thisRef: Any?, property: KProperty<*>, value: T) {
            synchronized(lock) {
                super.setValue(thisRef, property, value)
            }
        }
    }
```
Where plain `vetoable` is:
```
public inline fun <T> vetoable(initialValue: T, crossinline onChange: (property: KProperty<*>, oldValue: T, newValue: T) -> Boolean):
            ReadWriteProperty<Any?, T> =
        object : ObservableProperty<T>(initialValue) {
            override fun beforeChange(property: KProperty<*>, oldValue: T, newValue: T): Boolean = onChange(property, oldValue, newValue)
        }
```

Reply:
The problem with your proposed `safeVetoable` implementation is that the value `beforeChange` sees might not be the same value `setValue` sees, so there's still a race.

You'd probably need something that's a new implementation of `ObservableProperty` to get the lock scope in the right place.

Reply:
Can you elaborate on the scenario you're describing there? Want to make sure I understand it

Reply:
Current value is _n_.  Thread 1 wants to set to _n+1_; thread 2 wants to set to _n+2_.

On thread 1 we call `onChange(n, n+1)`, which returns `true`.  Thread then happens to get pre-empted.
On thread 2 we call `onChange(n, n+2)`, which returns `true`, and then `setValue` sets the value to _n+2_.
Thread 1 then resumes, calling `setValue` which sets the value back from _n+2_ to _n+1_.

Reply:
But the call to onChange happens within the context of the lock created in setValue.

Reply:
Oh! I see, I mis-understood the control flow. `super.setValue` is what actually calls `onChange`?